dev/buildtool/spin_commands.py

```diff
@@ -180,72 +176,116 @@ def __init__(self, factory, options, **kwargs):
-        major=match.group(1), minor=match.group(2), patch=match.group(3))
+    self.__spinnaker_version = options.bom_version or bom_contents['version']
+    self.__gate_version = gate_entry['version']
+    self.__stable_version = None # Set after promote_spin.
+    self.__no_changes = False # Set after promote_spin.
     self.__scm = BranchSourceCodeManager(options_copy, self.get_input_dir())
-    # TODO(jacobkiefer): Add spin CLI autogenerated docs.
-
-    dash = spinnaker_version.find('-')
-    semver_str = spinnaker_version[0:dash]
-    semver_parts = semver_str.split('.')
-    if len(semver_parts) != 3:
-      raise_and_log_error(
-          ConfigError('Expected spinnaker version in the form X.Y.Z-N'))
-    self.__release_branch = 'release-{maj}.{min}.x'.format(
-        maj=semver_parts[0], min=semver_parts[1])
-    self.__release_tag = 'version-' + semver_str
-
-    self.__release_version = semver_str
     self.__gcs_uploader = SpinGcsUploader(options)
 
+
   def push_tag_and_branch(self, repository):
     """Pushes a stable branch and git version tag to the origin repository."""
+    if self.__no_changes:
+      logging.info('No changes in spin since last tag, skipping branch and tag push.')
+      return
+
     git_dir = repository.git_dir
     git = self.__scm.git
 
-    release_url = repository.origin
-    logging.info('Pushing branch=%s and tag=%s to %s',
-                 self.__release_branch, self.__release_tag, release_url)
-
-    existing_commit = git.query_commit_at_tag(git_dir, self.__release_tag)
-    if existing_commit:
-      want_commit = git.query_local_repository_commit_id(git_dir)
-      if want_commit == existing_commit:
-        logging.debug('Already have "%s" at %s',
-                      self.__release_tag, want_commit)
-        return
+    match = re.match(r'(\d+)\.(\d+)\.(\d+)-\d+', self.__gate_version)
+    if match is None:
+      raise_and_log_error(
+          ConfigError('gate version {version} is not X.Y.Z-<buildnum>'
+                      .format(version=self.__gate_version)))
+    semver_parts = self.__spinnaker_version.split('.')
+    if len(semver_parts) != 3:
+      raise_and_log_error(
+          ConfigError('Expected spinnaker version in the form X.Y.Z-N, got {}'
+                      .format(self.__spinnaker_version)))
 
+    release_branch = 'release-{maj}.{min}.x'.format(
+        maj=semver_parts[0], min=semver_parts[1])
+    release_tag = 'version-' + self.__stable_version
+    logging.info('Pushing branch=%s and tag=%s to %s',
+                 release_branch, release_tag, repository.origin)
     git.check_run_sequence(
         git_dir,
         [
-            'checkout -b ' + self.__release_branch,
-            'remote add release ' + release_url,
-            'push release ' + self.__release_branch,
-            'tag ' + self.__release_tag,
-            'push release ' + self.__release_tag
+            'checkout -b ' + release_branch,
```

Comment:
I was confused earlier in this git UI that doesnt show much context and assumed this was the publisher class. Why does a build create a branch and tag? Builds should not have side effects into the repo.  I dont know why push and promote are in this class. Those are publisher activities.

Reply:
This is in `class PublishSpinCommand`.